File: runtime/go/units/wasi/write_string_list.go
// requires: memory/i32_store, memory/i32_store8, memory/init
func (w *WASI) write_string_list(strings [][]byte, listPtr, bufPtr uint32) uint32 {
    for i, s := range strings {
        w.memory.i32_store(uint64(listPtr)+uint64(i)*4, bufPtr)
        w.memory.init(uint64(bufPtr), s, 0, uint64(len(s)))
        w.memory.i32_store8(uint64(bufPtr)+uint64(len(s)), 0)
        bufPtr += uint32(len(s)) + 1
    }
    return wasiOk
}

File: runtime/java/units/wasi/write_string_list.java
// requires: memory/i32_store, memory/i32_store8, memory/init
// The WASI two-part layout shared by args_get/environ_get: a pointer array plus
// NUL-terminated bytes in a separate buffer.
int write_string_list(byte[][] xs, int listPtr, int bufPtr) {
    long lp = Integer.toUnsignedLong(listPtr);
    long bp = Integer.toUnsignedLong(bufPtr);
    for (byte[] s : xs) {
        memory.i32_store(lp, (int) bp);
        memory.init(bp, s, 0, s.length);
        memory.i32_store8(bp + s.length, 0);
        lp += 4;
        bp += s.length + 1;
    }
    return WASI_OK;
}
